File: Arrays/LargestNumber.py
# ──────────────────────────────────────────────────────────────────────
# LeetCode #179 · Largest Number
# Difficulty : Medium
# Topics     : Array, String, Greedy, Sorting
# URL        : https://leetcode.com/problems/largest-number/
# ──────────────────────────────────────────────────────────────────────
# Approach
#   The problem is to arrange a list of non-negative integers to form the
#   largest possible number. The key insight is to use a custom comparator
#   for sorting: for two strings a and b, we compare a+b vs b+a. If a+b >
#   b+a, then a should come before b. After sorting, if the first element
#   is "0", the entire number is zero. Otherwise, we join the sorted
#   strings. This greedy sorting approach works because the comparator
#   ensures the optimal order.
# 
# Complexity
#   Time  : O(n log n)
#   Space : O(n)
# 
# Runtime  : 8
# Memory   : 44964000
# 
# Examples
# 
# Constraints
# ──────────────────────────────────────────────────────────────────────

import pandas as pd

class Solution:
    def largestNumber(self, nums: List[int]) -> str:
        df = pd.DataFrame({'num': nums})
        df['str'] = df['num'].astype(str)
        # Custom comparator using sort_values with key is not straightforward; use Python's sort
        strs = df['str'].tolist()
        strs.sort(key=functools.cmp_to_key(lambda a, b: -1 if a + b > b + a else (1 if a + b < b + a else 0)))
        result = ''.join(strs)
        return '0' if result[0] == '0' else result

File: Strings/LargestNumber.go
/*
LeetCode Problem 179: Largest Number
Problem Number: 179
Difficulty: Medium
Link: https://leetcode.com/problems/largest-number/

Given a list of non-negative integers nums, arrange them such that they form the largest number.

Example 1:
Input: nums = [10,2]
Output: "210"

Example 2:
Input: nums = [3,30,34,5,9]
Output: "9534330"

Constraints:
- 1 <= nums.length <= 100
- 0 <= nums[i] <= 10^9

Topics: String, Sorting, Greedy
Time Complexity: O(n log n) - for sorting
Space Complexity: O(n) - for the string array
*/

package strings

import (
    "sort"
    "strconv"
    "strings"
)

func largestNumber(nums []int) string {
    // Convert numbers to strings
    numStrs := make([]string, len(nums))
    for i, num := range nums {
        numStrs[i] = strconv.Itoa(num)
    }
    
    // Sort using custom comparator: compare concatenated results
    sort.Slice(numStrs, func(i, j int) bool {
        return numStrs[i]+numStrs[j] > numStrs[j]+numStrs[i]
    })
    
    // If the largest number is "0", the result should be "0"
    if numStrs[0] == "0" {
        return "0"
    }
    
    // Build the result
    return strings.Join(numStrs, "")
}
